### PR TITLE
Mono: switch to CMake build on Windows

### PR DESCRIPTION
**Pull request by akoeplinger**
This unifies the build with the other platforms.

Hooks up the versioning targets so the mono libraries get the correct version.

### CODE REVIEW

**Comment by jkoritzinsky** [comment]
For consistency with the rest of the repo build, can we condition the switch to CMake+Ninja on the `$(Ninja)` property in MSBuild being set to `true`?

**Comment by akoeplinger** [comment]
@jkoritzinsky yep, that's what I'm already doing :)

File: src/mono/CMakeLists.txt [hunk 1]
```diff
@@ -206,31 +210,15 @@ elseif(CMAKE_SYSTEM_NAME STREQUAL "iOS" OR CMAKE_SYSTEM_NAME STREQUAL "tvOS")
...
   set(ENABLE_MONOTOUCH 1)
   add_definitions(-DMONOTOUCH=1)
-  set(HAVE_SYSTEM 0)
-  if(CMAKE_SYSTEM_NAME STREQUAL "tvOS")
-    set(HOST_TVOS 1)
-  endif()
-
-  # Force some defines
-  set(HAVE_GETPWUID_R 0)
-  set(HAVE_SYS_USER_H 0)
```
Comment (vargaz) [inline src/mono/CMakeLists.txt:222]: Why are these removed ? They were in the configure version probably because they are detected to be present, but don't work/shouldn't be used.

File: src/mono/CMakeLists.txt [hunk 2]
```diff
@@ -511,7 +529,7 @@ set(BREAKPOINT "G_STMT_START { raise(SIGTRAP); } G_STMT_END")
 if(GCC)
   set(GNUC_UNUSED "__attribute__((__unused__))")
   set(GNUC_NORETURN "__attribute__((__noreturn__))")
-  if(CMAKE_SYSTEM_PROCESSOR MATCHES "x86|x86_64")
+  if(TARGET_AMD64 OR TARGET_X86)
     set(BREAKPOINT "G_STMT_START { __asm__(\"int \$03\"); } G_STMT_END")
```
Comment (vargaz) [inline src/mono/CMakeLists.txt:533]: This should use host.

File: src/mono/cmake/configure.cmake [hunk 3]
```diff
@@ -69,7 +65,7 @@ ac_check_funcs (
   gethrtime read_real_time gethostbyname gethostbyname2 getnameinfo getifaddrs if_nametoindex
   access inet_ntop Qp2getifaddrs)
 
-if (NOT DARWIN)
+if (NOT TARGET_OSX)
   ac_check_funcs (getentropy)
```
Comment (vargaz) [inline src/mono/cmake/configure.cmake:69]: This should use host.

**Comment by vargaz** [comment]
The cmake changes looks ok to me otherwise.

File: eng/native/init-compiler-and-cmake.cmd [hunk 4]
```diff
@@ -0,0 +1,89 @@
+@if not defined _echo @echo off
```
Comment (akoeplinger) [inline eng/native/init-compiler-and-cmake.cmd:1]: this file was extracted from https://github.com/dotnet/runtime/blob/master/src/libraries/Native/build-native.cmd and/or https://github.com/dotnet/runtime/blob/master/src/installer/corehost/build.cmd, we can unify these later on.

**Review by steveisok — APPROVED** [review]

**Comment by CoffeeFlux** [comment]
Can we get rid of (and stop mirroring) the old solution as part of this PR?

**Comment by akoeplinger** [comment]
@CoffeeFlux yes, I'll do that in a follow-up PR (it needs a small tweak in the mirror and that is cumbersome to do in sync with this PR).

**Comment by akoeplinger** [comment]
Merging since this PR unblocks other work dependent on it, but feel free to still review and I'll address it in follow-ups.
